src/jmh/java/software/amazon/cloudwatchlogs/emf/MetricsLoggerBenchmark.java

```diff
@@ -0,0 +1,345 @@
+     */
+    @Benchmark
+    @Warmup(time = 10)
+    @Measurement(time = 10)
+    public void measureFlushWith1000Threads() throws InterruptedException {
+        measureFlushWithNThreads(1000);
+    }
+
+    /**
+     * Execute all methods for 1000 times with 5 threads. no lock (need to sync getAllDimensions() &
+     * getAllDimensionKeys() in MetricsDirective): 84.041 ± 24.965 ms/op; RW lock: 264.439 ± 8.070
+     * ms/op; S lock: 264.630 ± 24.252 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    public void measureAllMethodsWith5Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(5);
+    }
+
+    /**
+     * Execute all methods for 1000 times with 10 threads. no lock (need to sync getAllDimensions()
+     * & getAllDimensionKeys() in MetricsDirective): 41.174 ± 6.084 ms/op; RW lock: 263.103 ± 15.141
+     * ms/op; S lock: 256.267 ± 30.922 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    public void measureAllMethodsWith10Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(10);
+    }
+
+    /**
+     * Execute all methods for 1000 times with 100 threads. no lock (need to sync getAllDimensions()
+     * & getAllDimensionKeys() in MetricsDirective): 35.779 ± 2.414 ms/op; RW lock: 315.340 ± 16.074
+     * ms/op; S lock: 288.459 ± 5.801 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    @Warmup(time = 10)
+    @Measurement(time = 10)
+    public void measureAllMethodsWith100Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(100);
+    }
+
+    /**
+     * Execute all methods for 1000 times with 500 threads. no lock (need to sync getAllDimensions()
+     * & getAllDimensionKeys() in MetricsDirective): 81.785 ± 11.616 ms/op; RW lock: 346.697 ±
+     * 51.133 ms/op; S lock: 368.981 ± 161.049 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    @Warmup(time = 10)
+    @Measurement(time = 10)
+    public void measureAllMethodsWith500Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(500);
+    }
+
+    /**
+     * Execute all methods for 1000 times with 1000 threads. no lock (need to sync
+     * getAllDimensions() & getAllDimensionKeys() in MetricsDirective): 218.505 ± 178.808 ms/op; RW
+     * lock: 436.380 ± 317.130 ms/op; S lock: 390.074 ± 100 ms/op
+     *
+     * @throws InterruptedException
+     */
+    @Benchmark
+    @Warmup(time = 10)
+    @Measurement(time = 10)
+    public void measureAllMethodsWith1000Threads() throws InterruptedException {
+        measureAllMethodsWithNThreads(1000);
+    }
+
+    private void measurePutMetricWithNThreads(int n) throws InterruptedException {
+        logger = new MetricsLogger(envProvider);
+        Thread[] threads = new Thread[n];
+
+        for (int i = 0; i < n; i++) {
+            final int id = i;
+            int batchSize = 10000 / n;
```

Comment:
[nit] put this outside of for loop : )

Reply:
why don't we have a fixed number of batchSize regardless of input `n` ?
Imagine n=1000, then we will only have 10 batches, then each thread only execute 10 loops which is fast. 1000 threads will more likely not run at same time. Larger N, less batchSize, less time spent per thread, less concurrency